File: spell/spell.c
/**
 * A spelling corrector written in C which uses tries.
 *
 * To compile and run do: 
 *   gcc spell.c trie.c
 *   wget http://norvig.com/big.txt
 *   ./a.out speling < big.txt
 *
 * See http://norvig.com/spell-correct.html 
 *
 * Steve A <iradik@gmail.com>
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "map.h"

/* Note: Biggest word in big.txt is 18 characters. */
#define WORD_LEN_MAX 1024

void fatal(char *error) {
    printf("Fatal error %s\n", error);
    exit(-1); 
}

void scopy(char *dest, const char* src, int len) {
   strncpy(dest, src, len);
   dest[len]='\0';
}

void find_max(char* word, void* arg1, void* arg2, void* arg3) {
    result_t* found;
    map_t* map = (map_t*)arg1;
    int* max = (int*) arg2;
    char*max_word = (char*)arg3;

    if (found = find_map(map, word)) {
       if (found->count > *max) {
           *max = found->count;
           scopy(max_word, word, strlen(word));
       }
       //printf("%s\n", word);
    }
}

/** Counts words from STDIN into the map. */ 
void train(map_t *map) {
    static char word[WORD_LEN_MAX];
    char c;
    int i = 0;
    while ((c = tolower(getchar())) != EOF) {
       if (i > WORD_LEN_MAX) goto err;
       if (c < 'a' || c > 'z') {
               word[i] = '\0';
               load_map(map,word);
               i = 0;
       } else word[i++] = c;
    }
    return;
    err: fatal("WORD_LEN_MAX exceeded!");
}

void calc_edits1(const char *word, map_t* edits, void load_f(map_t*,char*)) {
   int word_len, i;
   char c;
   word_len = strlen(word);
   
   char *temp = malloc(word_len+3 * sizeof(char));
   char split[2][WORD_LEN_MAX];

   for (i=0; i < word_len+1; ++i) {
       /** split */
       scopy(split[0], word, i);
       scopy(split[1], &word[i], word_len - i);
       /** delete */
       if (i < word_len) {
           scopy(temp, split[0], i);
           scopy(&temp[i], &split[1][1], word_len - i - 1);
           load_f(edits, temp);
       }
       /** inserts */
       if (i < word_len+1) {
           for (c='a'; c <= 'z'; ++c) {
              scopy(temp, split[0], i);
              temp[i] = c;
              scopy(&temp[i+1], split[1], word_len - i);
              load_f(edits, temp);
           }
       }
       /** replaces */
       if (i < word_len) {
           for (c='a'; c <= 'z'; ++c) {
               scopy(temp, split[0], i);
               temp[i] = c;
               scopy(&temp[i+1], &split[1][1], word_len - i - 1);
               load_f(edits, temp);
           } 
       }
       /** transpose */
       if (i < word_len-1) {
           scopy(temp, split[0], i);
           temp[i] = split[1][1]; 
           temp[i+1] = split[1][0]; 
           scopy(&temp[i+2], &split[1][2], word_len - i - 2);
           load_f(edits, temp);
       }
   }

   free(temp);
}

static map_t* s_known_words;

void known_load_map(map_t* map, char *word) {
    if (find_map(s_known_words, word)) {
        load_map(map, word); 
    }    
}

void calc_edits2(char* edit1_word, void* arg1, void* arg2, void* arg3) {
    map_t* edits2 = (map_t*)arg1;
    calc_edits1(edit1_word, edits2, known_load_map);
}

void correct(map_t* known_words, char *word) {
   map_t *edits, *edits2;

   if (find_map(known_words, word)) {
       printf("%s\n", word);
       return;
   } 

   edits = create_map();
   calc_edits1(word, edits, load_map);

   int max = 0;
   char max_word[WORD_LEN_MAX];
  
   traverse_map(edits, 0, "", find_max, known_words, &max, max_word);

   if (!max) {
       edits2 = create_map();
       traverse_map(edits, 0, "", calc_edits2, edits2, NULL, NULL);
       traverse_map(edits2, 0, "", find_max, known_words, &max, max_word);
       delete_map(edits2);
   }
   delete_map(edits);

   printf("%s %d\n", max_word, max);   
}

int main(int argc, char **argv) {
   map_t *known_words = create_map();
   train(known_words);
   s_known_words = known_words;

   char *word = argv[1];
   correct(known_words, word);

   delete_map(known_words);
}

File: spell.go/spell.go
/**
 * "How to Write a Spelling Corrector" in Go
 * 
 * See original at http://norvig.com/spell-correct.html
 *
 * To compile, link, and run do:
 *   6g spell.go && 6l spell.6 && time ./6.out spellin < big.txt   
 */

package main

import (
    "os"
    "fmt"
    "bufio"
    "unicode"
)

func train() map[string] int {
    var word[1024] byte
    i := 0
    words := map[string] int {}
    reader := bufio.NewReader(os.Stdin)
    for {
       c,_ := reader.ReadByte() 
       c = byte(unicode.ToLower(int(c)))
       if c == 0 { break }
       if c  < 'a' || c > 'z' {
            words[string(word[0:i])]++
            i=0
        } else {
            word[i] = c
            i++
        }
    }
    return words
}

func calc_edits1(word string) []string {
    splits := [][]string{}
    for i := 0; i <= len(word); i++ {
        split := []string{word[0:i],word[i:]}
        splits = append(splits, split)
    }
   
    deletes := []string{} 
    for _,s := range splits {
        if len(s[1]) > 0 {
            deletes = append(deletes, s[0] + s[1][1:])
        }
    }

    replaces := []string{} 
    for _,s := range splits {
        if len(s[1]) > 0 {
            for _,c := range "abcdefghjkilmnopqrstuvwxyz" {
                replaces = append(replaces, s[0] + string(c) + s[1][1:])
            }
        }
    }

    transposes := []string{}
    for _,s := range splits {
        if len(s[1]) > 1 {
            transpose := s[0] + s[1][1:2] + s[1][0:1] + s[1][2:]
            transposes = append(transposes, transpose)
        }
    }

    inserts := []string{}
    for _,s := range splits {
        for _,c := range "abcdefghjkilmnopqrstuvwxyz" {
            insert := s[0] + string(c) + s[1]
            inserts = append(inserts, insert)
        }
    }

    all := []string{}
    all = append(all, deletes...)
    all = append(all, replaces...)
    all = append(all, transposes...)
    all = append(all, inserts...)
 
    return all
}

func known_words(nwords map[string] int, words []string) []string {
    known_words := []string{}
    for _,word := range words {
        if nwords[word] != 0 {
            known_words = append(known_words, word)
        } 
    }
    return known_words
}

func findmax(nwords map[string] int, kwords []string) string {
    max := 0
    max_word := ""
    for _,kword := range kwords {
        if (nwords[kword] > max) {
            max = nwords[kword]
            max_word = kword
        }
    }
    return max_word
}

func correct(nwords map[string] int, word string) string {
    kwords := known_words(nwords, []string {word})
    var edits1 []string
    if len(kwords) == 0 {
        edits1 = calc_edits1(word)
        kwords = known_words(nwords, edits1)
    }
    if len(kwords) == 0 {
        for _,edit1 := range edits1 {
            edits2 := calc_edits1(edit1)
            kwords = append(kwords, known_words(nwords, edits2)...)
        }
    }
    return findmax(nwords, kwords)
}

func usage() {
    fmt.Println("Usage: " + os.Args[0] + " word < big.txt")
}

func main() {
    if len(os.Args) < 2 { 
        usage()
        return;
    }
    nwords := train()
    correction := correct(nwords, os.Args[1])
    fmt.Println(correction)
}
